proposals/2025-03-20-URIs-for-Tags.md

```diff
@@ -0,0 +1,138 @@
+This proposal adds such an alternative, giving tags the same capabilities as all other similar mechanisms within the OAS.
+
+## Motivation
+
+### A user request and proposal
+
+From @SandroG in issue #2905 (which is only closed because it was consolidated into #3853), which got two further thumbs-ups:
+
+_**[NOTE:** The mechanism proposed here is **not** the one favored by this proposal, which is explained further down]_
+
+> I have a large specification, which I need to break down in different files. I use an approach where each file is like a sub-specification that lists all endpoints regarding the same subject and then I include these endpoints in the main openapi file.
+> The documentation of a tag is in the same file as the endpoint that uses it.
+> 
+> I'm not able to reuse that tag declaration in the main file, so I'm not able to include the description of the tag.
+> 
+> For example, I have a separate file for customer's endpoints
+
+`customers.yaml`
+```YAML
+info:
+...
+tags:
+  - name: Customer
+    description: APIs to manage customers. A customer is a representation of ...
+  
+paths:
+  /customers/{id}/:
+    parameters:
+      - name: id
+        . . .
+    get:
+      . . .
+  /customers/:
+    . . .
+```
+> I need to do this:
+
+`openapi.yaml`
+```YAML
+info:
+...
+tags:
+  - $ref: "./customers.yaml#/tags/0"
+  
+paths:
+  /customers/{id}/:
+    $ref: "./customers.yaml#/paths/~customers~1{id}~1"
+  /customers/:
+    $ref: "./customers.yaml#/paths/~customers~1"
+```
+
+### Proof of confusion over tag to Tag Object resolution
+
+In the above example, @SandroG proposes using `$ref` for Tag Objects to pull Tag Objects from one [OpenAPI Document](https://spec.openapis.org/oas/v3.1.1#openapi-document) into another, where they can be used by the `tags` field in Operation Objects.
+This makes clear that they expect tags to be resolved from the _current document_, which may be a [referenced document rather than the entry document](https://spec.openapis.org/oas/v3.1.1#openapi-description).
+
+However, in [3.0.4](https://spec.openapis.org/oas/v3.0.4#resolving-implicit-connections) and [3.1.1](https://spec.openapis.org/oas/v3.1.1#resolving-implicit-connections) we RECOMMEND (== SHOULD) that tag names are resolved to Tag Objects in the _entry document_.
+This means that there is no way to resolve them from the current document, which is the mirror image of the problem as that encountered by @SandroG.
+
+In today's TDC call, @lornajane stated that she expects tag names to be resolved from the entry document, and @kevinswiber expressed doubt that anyone implements anything else (sadly @SandroG does not mention their tool, which presumably resolves from the current document or else they would not have explained the issue in this way).
```

Comment:
I'm sorry I missed this discussion in the TDC call. I can easily imagine scenarios where tags should be resolved to the current document. Consider an API that is actually implemented as a collection of microservices that each implement the CRUD operations for one or more resources. The API description might be structured as an entry document that $ref's in groups of path items from separate documents for each microservice. Now suppose we want one tag per resource -- these tags would naturally be defined in the document for a micro service and not the entry document.

Reply:
@mikekistler yes, that's exactly the sort of scenario I feel we should support using the same mechanisms we use for everything else.

I'm a little curious why you see this scenario but reject solving it?

Reply:
@mikekistler got it.  yeah in 4.0 we need to have just one mechanism.  Or at least one mechanism per use case (e.g. one magic "look in the entry document" mechanism, and one explicit "look exactly where I say" mechanism).  Note that in JSON Schema, `$dynamicRef` is the (more flexible but also more complex) equivalent of the magic entry document mechanism.